src/core/bpf-pinned-prog.c

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "bpf-pinned-prog.h"
+#include "cgroup-bpf-ctx.h"
+
+#include "unit.h"
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(bpf_prog_hash_ops, void, trivial_hash_func, trivial_compare_func, BPFProgram, bpf_program_unref);
+
+static int load_programs(Unit *u, const CGroupContext *c, Hashmap *progs_by_id) {
+        CGroupPinnedBPFProgram *p = NULL;
+        int r;
+
+        assert(u);
+        assert(c);
+        assert(progs_by_id);
+
+        LIST_FOREACH(prog, p, c->pinned_bpf_progs) {
+                _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+
+                r = bpf_program_new_from_bpffs_path(p->bpffs_path, p->attach_type, &prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Loading of BPF program %s failed: %m", p->bpffs_path);
+
+                r = hashmap_put(progs_by_id, p, TAKE_PTR(prog));
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't put program to hashmap: %m");
+        }
+
+        return 0;
+}
+
+static int attach_programs(Unit *u, const char *path, Hashmap *progs_by_id, Set **set_installed, uint32_t attach_flags) {
+        CGroupPinnedBPFProgram *key;
+        BPFProgram *prog;
+        int r;
+
+        assert(u);
+
+        set_clear(*set_installed);
+
+        HASHMAP_FOREACH_KEY(prog, key, progs_by_id) {
+                r = bpf_program_cgroup_attach(prog, key->attach_type, path, attach_flags);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Attaching custom egress BPF program to cgroup %s failed: %m", path);
+
+                r = set_ensure_put(set_installed, &bpf_prog_hash_ops, prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't add program to BPF program set: %m");
+                bpf_program_ref(prog);
+        }
+
+        return 0;
+}
+
+int bpf_pinned_prog_supported(void) {
+        int r;
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EUCLEAN),
```

Comment:
@anitazha 

Switched to `cg_all_unified() > 0`.
Made `supported` for consistency with `bpf_{firewall|devices}_supported` which aside from checking for unified hierarchy also try to load a trivial program. Pinned programs are already loaded to the kernel, no need for that.
The check I might add is probing for attach flags (multi), which support was added in `v4.15` and bailing early in `support` method. But I don't have strong opinion whether the case worths special treatment since `4.15` is a pretty old kernel.